FZ: warn w/ a toast if an elevated window cannot be dragged

yuyoyuppe opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
- FZ: warn with a toast if an elevated window cannot be dragged. Warns once per PowerToys.exe launch, can be disabled for good using "Don't show again" button
- MSIX: update PowerRenameUWPUI path, so PowerRename works in MSIX again
- minor control-flow clean up

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [x] Applies to #1015

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
- As described in [that comment](https://github.com/microsoft/PowerToys/issues/1015#issuecomment-569241482), we cannot detect if the Shift key is being held while an elevated window is active due to [UIPI Windows security feature](https://en.wikipedia.org/wiki/User_Interface_Privilege_Isolation), so the toast notification is shown as soon as we drag an elevated window even w/o actually holding the Shift key.
- We need to store "Don't show again" flag state somewhere. UWP API offers [a convenient key-value storage](https://docs.microsoft.com/en-us/uwp/api/windows.storage.applicationdata.localsettings), but unfortunately it's not usable for unpackaged apps (throws "The process has no package identity."), so we use the same mechanism as [SHMessageBoxCheck](https://docs.microsoft.com/en-au/windows/win32/api/shlwapi/nf-shlwapi-shmessageboxchecka?redirectedfrom=MSDN) - the registry.
- Since we use the registry, we must introduce a custom URI scheme protocol which will be handled by a PowerToys.exe, because `notifications` library, being C++/WinRT project, doesn't support any legacy APIs such registry manipulation, so we can't use a background handler to update the flag state.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
- MSI works as admin/nonadmin (i.e. toast isn't shown when the window is draggable)
- MSIX works (via virtualized registry)

## Screenshot
![image](https://user-images.githubusercontent.com/1828123/77435942-f0a44780-6df3-11ea-83fa-3d12ecb210c0.png)

yuyoyuppe committed:
FZ: warn w/ a toast if an elevated window cannot be dragged and offer learning more

enricogior commented:
The message text needs some re-wording and we should not use `PT` for `PowerToys`.

yuyoyuppe commented:
What wording should we use? @enricogior @crutkas

crutkas commented:
We’ve detected an application launched as admin.  This will block some functionality in PowerToys.  Relaunch as Admin?

Thoughts as Context menu?
https://docs.microsoft.com/en-us/windows/uwp/design/shell/tiles-and-notifications/adaptive-interactive-toasts
“Yes”, “No”, “Learn More”, “Don’t Show for 24hrs”
Why I like context menu, it makes it harder to click the “don’t show again” and provides a learning moment too.

'Yes' -> relaunch as admin
'No'-> ignore
'Learn more' -> link to github docs page
'Don’t Show for 24hrs' -> don't prompt for 24hrs

We can also have an option in settings to "Notify me when apps when applications are launched in a state where PowerToys functionality would be inhibited.  Typically this is when they are run as administrator"

yuyoyuppe commented:
@crutkas
I might be missing something, but having more than 3 buttons on a toast clips the text content, and we can't `style` the `actions`:

![image](https://user-images.githubusercontent.com/1828123/77476534-cec6b700-6e2b-11ea-9bcf-b286e26b6301.png)

So, IIUC, you'd like us to have only 2 buttons instead and 2 context menu actions like this:

![image](https://user-images.githubusercontent.com/1828123/77476857-76dc8000-6e2c-11ea-96d4-a1657752237f.png)

Or do we have a third button which snoozes a notification for 24 hrs and a context menu action which disables it for good:

![image](https://user-images.githubusercontent.com/1828123/77477358-421cf880-6e2d-11ea-9be9-60a743a2fc4d.png)

> Thoughts as Context menu?

I'm ok with it, my only concern is not to produce too many notifications in case a user isn't aware of the toast context menu functionality and is forced to snooze a toast every day.

By the way, I'm using [Notifications Visualizer](https://www.microsoft.com/en-us/p/notifications-visualizer/9nblggh5xsl1?activetab=pivot:overviewtab) to produce these toast sketches, here's an example markup:

<details><summary>3 buttons toast markup</summary>
<p>

```xml
<toast>
  
  <visual>
    <binding template="ToastGeneric">
      <text>PowerToys</text>
      <text>We’ve detected an application launched as admin. This will block some functionality in PowerToys. Relaunch as Admin?</text>
    </binding>
  </visual>
<actions>
    <action
      hint-weight="1"
      activationType="background"
      arguments="snooze"
      content="Yes"/>
    <action
      activationType="background"
      arguments="snooze"
      content="No"/>
    <action
      activationType="background"
      arguments="snooze"
      content="Snooze"/>

    <action
      activationType="background"
      placement="contextMenu"
      arguments="learning"
      content="Learn More"/>
    <action
      activationType="background"
      placement="contextMenu"
      arguments="waiting"
      content="Don’t Show Again"/>
</actions>
  
</toast>
```
</p>
</details>

enricogior commented on installer/MSIX/PackagingLayout.xml:
```diff
@@ -16,7 +16,7 @@
         <File DestinationPath="modules\PowerRenameExt.dll" SourcePath="..\..\x64\Release\modules\PowerRenameExt.dll"/>
         <File DestinationPath="modules\shortcut_guide.dll" SourcePath="..\..\x64\Release\modules\shortcut_guide.dll"/>
-        <File DestinationPath="modules\PowerRenameUWPUI.exe" SourcePath="..\..\x64\Release\PowerRenameUWPUI.exe"/>
+        <File DestinationPath="modules\PowerRenameUWPUI.exe" SourcePath="..\..\x64\Release\modules\PowerRenameUWPUI.exe"/>
```
Is this fix related to this PR or was something missing in a previous commit?

yuyoyuppe replied:
unrelated, wanna make a separate PR?

yuyoyuppe replied:
unrelated, I'll make a separate PR

enricogior commented on src/modules/fancyzones/lib/FancyZones.cpp:
```diff
@@ -954,54 +973,56 @@ void FancyZones::MoveSizeEndInternal(HWND window, POINT const& ptScreen, require
 void FancyZones::MoveSizeUpdateInternal(HMONITOR monitor, POINT const& ptScreen, require_write_lock writeLock) noexcept
 {
-    if (m_inMoveSize)
+    if (!m_inMoveSize)
```
The only change in MoveSizeUpdateInternal is the inverted logic?
In that case it would be better to do these type of changes in their own PRs.

yuyoyuppe replied:
sure, I'll create a separate PR

enricogior commented:
We don't need the "No" button on a notification, since notifications are not supposed to be dismissed with a button.

crutkas commented:
> We don't need the "No" button on a notification, since notifications are not supposed to be dismissed with a button.

still prefer the idea to not perm hid it, too easy via a button to misclick it IMO.  Hence the combo

yuyoyuppe commented:
> We don't need the "No" button on a notification, since notifications are not supposed to be dismissed with a button.

It supported by having an empty handler.

Alright, I'll go with Yes/Snooze and Learn More/Don't show again context menu items.

enricogior commented:
@crutkas 
it turned out that relaunching as admin may not be the best UX unless we explain to the user that:
 - it's a one time restart as admin, next time PT will restart as user
 - to make it permanent the user has to select the option in the Settings
 - in theory we can change that option silently, but do we? I think it's not ideal because first of all we don't make that clear to the user, and doing so basically we defeat the whole idea of trying to not run elevated all the time unless the user really wants to do so.
 - if we decide to make the option of running as admin permanent, we also have to recreate the scheduled task if the `Start at login` option is on and that requires to run already elevated, so it's doable but not completely trivial to do it in this work item just before releasing 0.16.

Proposed alternative: what if we just offer to the users go to the Settings so they decide themselves what to turn on? We can open the Settings for them if they click yes on the notification.

Nit: saying "We’ve detected an application launched as admin" may be confusing since there are several system apps (Task Manager, Computer Manager, etc.) that run elevated without having been launched as administrator. I understand that we can't be too verbose, but the correct description would be something like "an application running with administrator privileges".

yuyoyuppe commented:
> in theory we can change that option silently, but do we? I think it's not ideal because first of all we don't make that clear to the user, and doing so basically we defeat the whole idea of trying to not run elevated all the time unless the user really wants to do so.

Regarding this, we could also have a third button like "Yes, always"/"Always".

yuyoyuppe commented:
(not ready for review yet)

yuyoyuppe commented:
@enricogior I've implemented your proposal to have a toast button for openning the settings window, but the window is opened as a background window, so I've gone ahead and implemented the "Yes, always" button:

![Untitled](https://user-images.githubusercontent.com/1828123/77639940-9974b300-6f6a-11ea-8fb6-49f4cd23da57.png)

We must use always have a snooze interval dropdown, therefore the "24 hours" thing.
If that change is deemed too risky, we can always rollback to a simple "non-actionable" notification.

yuyoyuppe commented:
- Updated to only offer a "Learn more" button, which redirects to the help page (doesn't exist yet).

![image](https://user-images.githubusercontent.com/1828123/77680974-c0040f80-6fa5-11ea-9c63-36f896a394ef.png)

- "Don't show again" will disable the notification for 30 days using a registry key

yuyoyuppe committed:
settings: add "Learn more about Admin mode" link

yuyoyuppe commented:
Added a "Learn more about Admin mode" link as a separate commit under the settings page:
![image](https://user-images.githubusercontent.com/1828123/77685613-bd58e880-6fac-11ea-8ebe-40cb2e1e2648.png)

enricogior reviewed (approved):
Great work!
